src/tdsops.f90

```diff
@@ -0,0 +1,700 @@
+      real(dp) :: alpha, delta
+   contains
+      procedure :: deriv_1st, deriv_2nd, interpl_mid, stag_der_1st, preprocess
+   end type tdsops_t
+
+   interface tdsops_t
+      module procedure tdsops
+   end interface tdsops_t
+
+contains
+
+   type(tdsops_t) function tdsops(n, delta, operation, variation, n_halo, &
+                                  from_to, bc_top, bc_bot, paire, c_nu, nu0_nu)
+      !! Constructor function for the tdsops_t class.
+      !! 
+      !! 'n', 'delta', 'operation', and 'variation' are necessary arguments.
+      !! Number of points 'n', distance between two points 'delta', the
+      !! 'operation' the tridiagonal system defines, and the 'variation' that
+      !! specifies the exact scheme we choose to apply for the operation.
+      !! The remaining arguments are optional.
+      !! 'from_to' is necessary for interpolation and staggared derivative, and
+      !! it can be 'v2p' or 'p2v'.
+      !! If the specific region the instance is operating is not a boundary
+      !! region, then 'bc_top' and 'bc_bot' are either 'null' or not provided.
+      !! 'sym' is relevant when the boundary condition is free-slip. If sym is
+      !! .true. then it means the field we operate on is assumed to be an even
+      !! function (symmetric) accross the boundary. If it is .false. it means
+      !! that the field is assumed to be an odd function (anti-symmetric).
+      !! 'c_nu', 'nu0_nu' are relevant when operation is second order
+      !! derivative and variation is compact6-hyperviscous.
+      implicit none
+
+      integer, intent(in) :: n
+      real(dp), intent(in) :: delta
+      character(*), intent(in) :: operation, variation
+      integer, optional, intent(in) :: n_halo
+      character(*), optional, intent(in) :: from_to, bc_top, bc_bot
+      logical, optional, intent(in) :: paire
+      real(dp), optional, intent(in) :: c_nu, nu0_nu
+
+      integer :: n_stencil
+
+      tdsops%n = n
+      tdsops%delta = delta
+      if (present(n_halo)) then
+         tdsops%n_halo = n_halo
+      else
+         tdsops%n_halo = 4
+      end if
+
+      n_stencil = 2*tdsops%n_halo + 1
+
+      allocate(tdsops%dist_fw(n), tdsops%dist_bw(n))
+      allocate(tdsops%dist_sa(n), tdsops%dist_sc(n))
+      allocate(tdsops%coeffs(n_stencil))
+      allocate(tdsops%coeffs_t(tdsops%n_halo, n_stencil))
+      allocate(tdsops%coeffs_b(tdsops%n_halo, n_stencil))
+
+      if (operation == 'first-deriv') then
+         call tdsops%deriv_1st(variation, bc_top, bc_bot, paire)
+      else if (operation == 'second-deriv') then
+         call tdsops%deriv_2nd(variation, bc_top, bc_bot, paire, c_nu, nu0_nu)
+      else if (operation == 'interpolate') then
+         call tdsops%interpl_mid(variation, from_to, bc_top, bc_bot, paire)
+      else if (operation == 'staggared-deriv') then
+         call tdsops%stag_der_1st(variation, from_to, bc_top, bc_bot, paire)
+      else
+         error stop 'operation is not defined'
```

Comment:
I think it would be nice to have a proper warning / error catching mechanism. See the one in decomp2d maybe ? Similarly, a proper log mechanism is probably needed (output to file or stdout ? only master rank will output or all ranks will output ?)

I think it is good to plan this a bit with some verbose / debug flags or variables.

However, this is beyond the scope of the present PR.

Reply:
Can you point me to this mechanism in decomp2d? A simple example if possible

Reply:
We use `decomp_2d_abort` and `decomp_2d_warning` to deal with errors and warnings.

The function is defined here : https://github.com/2decomp-fft/2decomp-fft/blob/dev/src/decomp_2d_mpi.f90#L133

It is used here for instance : https://github.com/2decomp-fft/2decomp-fft/blob/dev/src/io.f90#L128

For the log we have the following module. I think it can be improved, but it is not so bad : https://github.com/2decomp-fft/2decomp-fft/blob/dev/src/log.f90

Reply:
We decided to have unit tests as a error catching mechanism for the moment.